Added parameter validation for CountVectorizer & TfidfVectorizer

nymarkkho opened the pull request:
Reference Issues/PRs
See https://github.com/scikit-learn/scikit-learn/issues/23462

What does this implement/fix? Explain your changes.
Add validators for CountVectorizer. Towards https://github.com/scikit-learn/scikit-learn/issues/23462

- Defines _parameter_constraints in CountVectorizer.
- Removed simple validation in countvectorizer

Any other comments?
I need help in an error I'm receiving when I run `pytest -vl sklearn/tests/test_common.py -k check_param_validation`. The parameter validation is not raising an error message for the parameter `analyzer`

nymarkkho committed:
Parameter valdiation for CountVectorizer

nymarkkho committed:
Removal of simple validation for min_df, max_df and max_features

nymarkkho committed:
Transfer parameter validation to fit from fit_transform

Micky774 commented:
Okay so the major issue here is that `CountVectorizer` inherits from `_VectorizerMixin` which provides its own `_validate_params` method separate from that of `BaseEstimator`. The simplest route would be to rename it something else and make sure to call it after the initial param validation, since it is a complex validation (in the sense that it validates parameters against other parameters, albeit elements of the same tuple) and we're okay with those occurring alongside the main param validation API.

nymarkkho commented:
> Okay so the major issue here is that `CountVectorizer` inherits from `_VectorizerMixin` which provides its own `_validate_params` method separate from that of `BaseEstimator`. The simplest route would be to rename it something else (e.g. `_validate_ngram_range`) and make sure to call it after the initial param validation, since it is a complex validation (in the sense that it validates parameters against other parameters, albeit elements of the same tuple) and we're okay with those occurring alongside the main param validation API.

I've added that changes in the `_VectorizerMixin` and it still provides the same error for the initial parameter validation. From my understanding, the `_validate_params` in `_VectorizerMixin` validates the parameter `ngram_range` and would not affect the validation of `analyzer`.

Micky774 committed:
Renaming and moving validation calls

Micky774 commented:
> I've added that changes in the `_VectorizerMixin` and it still provides the same error for the initial parameter validation. From my understanding, the `_validate_params` in `_VectorizerMixin` validates the parameter `ngram_range` and would not affect the validation of `analyzer`.

I opened a [PR](https://github.com/nymarkkho/scikit-learn/pull/1) on your fork with a commit containing the changes I suggested. Note that even after the immediate problems are fixed, you next need to ensure that estimators which inherit from `CountVectorizer` and directly call its `fit` or `fit_transform` methods (e.g. `TfidfVectorizer`) also have the correct parameter constraints.

nymarkkho committed:
Merge pull request #1 from Micky774/param_validation_countvectorizer

Renaming and moving validation calls

nymarkkho committed:
Paraneter validation for TfidfVectorizer

nymarkkho committed:
Revert "Paraneter validation for TfidfVectorizer"

This reverts commit 73b059139cbb0b4a488107c7ce02b766d7df514c.

nymarkkho committed:
Parameter validation for TfidfVectorizer

nymarkkho committed:
Reformating

Micky774 commented:
@nymarkkho Can you address the failing tests? There's still some work to do on this PR. In particular you'll still need to remove additional tests that check the old validation strategy, e.g. `test_vectorizer_params_validation`. You can usually catch most of them by searching for `ValueError` or `TypeError`. Let me know if you have any questions or need any help :)

nymarkkho committed:
Removing simple tests

nymarkkho committed:
Adding parameter constraints required for tests

nymarkkho commented:
Thanks for the input. I've updated the PR which fixed the failing test. I've also added parameter constraints that weren't in the documentation to accommodate the test.

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/nymarkkho/23853

jeremiedbb committed:
docstring fixes and do not validate dtype

jeremiedbb reviewed (approved):
LGTM. Thanks @nymarkkho.

I just pushed a few fixes to the docstrings when all valid options were not properly listed
